syntax = "proto3";
package chat;

service ChatService {
  // Registers a new user.
  rpc Register(RegisterRequest) returns (StatusReply);
  // Logs in an existing user.
  rpc Login(LoginRequest) returns (LoginReply);
  // Sends a message from sender to recipient.
  rpc SendMessage(MessageRequest) returns (StatusReply);
  // Retrieves offline messages.
  rpc RetrieveOfflineMessages(OfflineRequest) returns (OfflineReply);
  // Retrieves message history.
  rpc MessageHistory(HistoryRequest) returns (HistoryReply);
  // Lists users matching a pattern.
  rpc ListUsers(UserListRequest) returns (UserListReply);
  // Deletes a message.
  rpc DeleteMessage(DeleteMessageRequest) returns (StatusReply);
  // Deletes a user account.
  rpc DeleteAccount(AccountRequest) returns (StatusReply);
  // (Optionally) signals quit/disconnect.
  rpc Quit(Empty) returns (StatusReply);
}

message RegisterRequest {
  string username = 1;
  string password = 2;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message MessageRequest {
  string sender = 1;
  string recipient = 2;
  string message = 3;
}

message OfflineRequest {
  string username = 1;
  int32 count = 2;
}

message OfflineReply {
  repeated MessageData messages = 1;
}

message HistoryRequest {
  string username = 1;
}

message HistoryReply {
  repeated MessageData messages = 1;
}

message UserListRequest {
  string pattern = 1;
}

message UserListReply {
  repeated string usernames = 1;
}

message DeleteMessageRequest {
  string username = 1;
  string message_id = 2;
}

message AccountRequest {
  string username = 1;
}

message MessageData {
  string id = 1;
  string sender = 2;
  string recipient = 3;
  string content = 4;
}

message LoginReply {
  bool success = 1;
  string error_message = 2;
}

message StatusReply {
  bool success = 1;
  string error_message = 2;
}

message Empty {}
